### PR TITLE
Add more context to ms rules / identity provider callbacks

### PR DESCRIPTION
**Pull request by mulmarta**
Maybe useful to know more context

### Call-outs

Moving GroupContext to core means making `tree_hash` and `confirmed_transcript_hash` public (there's currently a function on ExternalGroup that returns `confirmed_transcript_hash` anyway).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT license.

### CODE REVIEW

**Comment by tomleavy** [comment]
@mulmarta holding off on merging this for a week or so to see if we need any non-breaking changes first

File: mls-rs-core/src/identity/provider.rs
Comment (tomleavy) [inline mls-rs-core/src/identity/provider.rs:12]: Copy feels weird for this one but I guess it's ok?

File: mls-rs-core/src/lib.rs
Comment (tomleavy) [inline mls-rs-core/src/lib.rs:12]: We have a mod for group already, should this be a submodule of that?

File: mls-rs/src/group/message_processor.rs
Comment (tomleavy) [inline mls-rs/src/group/message_processor.rs:671]: Is the switch in order here so that if the update path doesn't validate we don't waste our time calculating the other stuff?

File: mls-rs/src/group/mod.rs
Comment (tomleavy) [inline mls-rs/src/group/mod.rs:336]: Feels like this should be some sort of constructor function given that we need epoch and confirmed_transcript_hash to be default values?

File: mls-rs/src/group/proposal_filter/filtering.rs
Comment (tomleavy) [inline mls-rs/src/group/proposal_filter/filtering.rs:180]: can we change this variable name, feels like it would be the original extensions, but it isn't

File: mls-rs/src/group/transcript_hash.rs
Comment (tomleavy) [inline mls-rs/src/group/transcript_hash.rs:37]: One thing we should decide ... maybe core can have things like this in there as well. Doesn't need to change now, but it feels like it might be useful for these small crypto operations that will never change to exist in core to keep things all toegether

**Review by tomleavy — COMMENTED** [review]
I really like the direction of this! Left some comments

File: mls-rs/src/group/message_processor.rs
Comment (mulmarta) [reply mls-rs/src/group/message_processor.rs:671]: That's a good reason but also we borrow `self` as immutable in `let group_state = self.group_state()` in line 626 so we can't call `self.group_state_mut()` to pop the reinit here.

File: mls-rs/src/group/transcript_hash.rs
Comment (mulmarta) [reply mls-rs/src/group/transcript_hash.rs:37]: Let's think about it. I'm not the greatest fan of the interim/confirmed transcript hash concept - one hash can be enough for security, and it could be computed more easily.

File: mls-rs/src/group/mod.rs
Comment (mulmarta) [reply mls-rs/src/group/mod.rs]: It only appears here, but added a constructor.

**Comment by mulmarta** [comment]
I fixed the merge conflict (version in Cargo.toml), I think I addressed all the comments, and we waited 2 weeks. OK to merge @tomleavy ?

**Review by tomleavy — APPROVED** [review]

**Comment by tomleavy** [comment]
> I fixed the merge conflict (version in Cargo.toml), I think I addressed all the comments, and we waited 2 weeks. OK to merge @tomleavy ?

Yup, I think we are good

**Review by stefunctional — APPROVED** [review]

**Pull request merged**
